File: RxTools-library/src/main/java/com/vondear/rxtools/view/likeview/tools/ei/RxEasingProvider.java
package com.vondear.rxtools.view.likeview.tools.ei;

import android.support.annotation.NonNull;

/**
 * The Easing class provides a collection of ease functions. It does not use the standard 4 param
 * ease signature. Instead it uses a single param which indicates the current linear ratio (0 to 1) of the tween.
 */
class RxEasingProvider {
    /**
     * @param rxEase            Easing type
     * @param elapsedTimeRate Elapsed time / Total time
     * @return easedValue
     */
    public static float get(@NonNull RxEase rxEase, float elapsedTimeRate) {
        switch (rxEase) {
            case LINEAR:
                return elapsedTimeRate;
            case QUAD_IN:
                return getPowIn(elapsedTimeRate, 2);
            case QUAD_OUT:
                return getPowOut(elapsedTimeRate, 2);
            case QUAD_IN_OUT:
                return getPowInOut(elapsedTimeRate, 2);
            case CUBIC_IN:
                return getPowIn(elapsedTimeRate, 3);
            case CUBIC_OUT:
                return getPowOut(elapsedTimeRate, 3);
            case CUBIC_IN_OUT:
                return getPowInOut(elapsedTimeRate, 3);
            case QUART_IN:
                return getPowIn(elapsedTimeRate, 4);
            case QUART_OUT:
                return getPowOut(elapsedTimeRate, 4);
            case QUART_IN_OUT:
                return getPowInOut(elapsedTimeRate, 4);
            case QUINT_IN:
                return getPowIn(elapsedTimeRate, 5);
            case QUINT_OUT:
                return getPowOut(elapsedTimeRate, 5);
            case QUINT_IN_OUT:
                return getPowInOut(elapsedTimeRate, 5);
            case SINE_IN:
                return (float) (1f - Math.cos(elapsedTimeRate * Math.PI / 2f));
            case SINE_OUT:
                return (float) Math.sin(elapsedTimeRate * Math.PI / 2f);
            case SINE_IN_OUT:
                return (float) (-0.5f * (Math.cos(Math.PI * elapsedTimeRate) - 1f));
            case BACK_IN:
                return (float) (elapsedTimeRate * elapsedTimeRate * ((1.7 + 1f) * elapsedTimeRate - 1.7));
            case BACK_OUT:
                return (float) (--elapsedTimeRate * elapsedTimeRate * ((1.7 + 1f) * elapsedTimeRate + 1.7) + 1f);
            case BACK_IN_OUT:
                return getBackInOut(elapsedTimeRate, 1.7f);
            case CIRC_IN:
                return (float) -(Math.sqrt(1f - elapsedTimeRate * elapsedTimeRate) - 1);
            case CIRC_OUT:
                return (float) Math.sqrt(1f - (--elapsedTimeRate) * elapsedTimeRate);
            case CIRC_IN_OUT:
                if ((elapsedTimeRate *= 2f) < 1f) {
                    return (float) (-0.5f * (Math.sqrt(1f - elapsedTimeRate * elapsedTimeRate) - 1f));
                }
                return (float) (0.5f * (Math.sqrt(1f - (elapsedTimeRate -= 2f) * elapsedTimeRate) + 1f));
            case BOUNCE_IN:
                return getBounceIn(elapsedTimeRate);
            case BOUNCE_OUT:
                return getBounceOut(elapsedTimeRate);
            case BOUNCE_IN_OUT:
                if (elapsedTimeRate < 0.5f) {
                    return getBounceIn(elapsedTimeRate * 2f) * 0.5f;
                }
                return getBounceOut(elapsedTimeRate * 2f - 1f) * 0.5f + 0.5f;
            case ELASTIC_IN:
                return getElasticIn(elapsedTimeRate, 1, 0.3);

            case ELASTIC_OUT:
                return getElasticOut(elapsedTimeRate, 1, 0.3);

            case ELASTIC_IN_OUT:
                return getElasticInOut(elapsedTimeRate, 1, 0.45);

            default:
                return elapsedTimeRate;

        }

    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @param pow             pow The exponent to use (ex. 3 would return a cubic ease).
     * @return easedValue
     */
    private static float getPowIn(float elapsedTimeRate, double pow) {
        return (float) Math.pow(elapsedTimeRate, pow);
    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @param pow             pow The exponent to use (ex. 3 would return a cubic ease).
     * @return easedValue
     */
    private static float getPowOut(float elapsedTimeRate, double pow) {
        return (float) ((float) 1 - Math.pow(1 - elapsedTimeRate, pow));
    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @param pow             pow The exponent to use (ex. 3 would return a cubic ease).
     * @return easedValue
     */
    private static float getPowInOut(float elapsedTimeRate, double pow) {
        if ((elapsedTimeRate *= 2) < 1) {
            return (float) (0.5 * Math.pow(elapsedTimeRate, pow));
        }

        return (float) (1 - 0.5 * Math.abs(Math.pow(2 - elapsedTimeRate, pow)));
    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @param amount          amount The strength of the ease.
     * @return easedValue
     */
    private static float getBackInOut(float elapsedTimeRate, float amount) {
        amount *= 1.525;
        if ((elapsedTimeRate *= 2) < 1) {
            return (float) (0.5 * (elapsedTimeRate * elapsedTimeRate * ((amount + 1) * elapsedTimeRate - amount)));
        }
        return (float) (0.5 * ((elapsedTimeRate -= 2) * elapsedTimeRate * ((amount + 1) * elapsedTimeRate + amount) + 2));
    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @return easedValue
     */
    private static float getBounceIn(float elapsedTimeRate) {
        return 1f - getBounceOut(1f - elapsedTimeRate);
    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @return easedValue
     */
    private static float getBounceOut(float elapsedTimeRate) {
        if (elapsedTimeRate < 1 / 2.75) {
            return (float) (7.5625 * elapsedTimeRate * elapsedTimeRate);
        } else if (elapsedTimeRate < 2 / 2.75) {
            return (float) (7.5625 * (elapsedTimeRate -= 1.5 / 2.75) * elapsedTimeRate + 0.75);
        } else if (elapsedTimeRate < 2.5 / 2.75) {
            return (float) (7.5625 * (elapsedTimeRate -= 2.25 / 2.75) * elapsedTimeRate + 0.9375);
        } else {
            return (float) (7.5625 * (elapsedTimeRate -= 2.625 / 2.75) * elapsedTimeRate + 0.984375);
        }
    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @param amplitude       Amplitude of easing
     * @param period          Animation of period
     * @return easedValue
     */
    private static float getElasticIn(float elapsedTimeRate, double amplitude, double period) {
        if (elapsedTimeRate == 0 || elapsedTimeRate == 1) return elapsedTimeRate;
        double pi2 = Math.PI * 2;
        double s = period / pi2 * Math.asin(1 / amplitude);
        return (float) -(amplitude * Math.pow(2f, 10f * (elapsedTimeRate -= 1f)) * Math.sin((elapsedTimeRate - s) * pi2 / period));
    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @param amplitude       Amplitude of easing
     * @param period          Animation of period
     * @return easedValue
     */
    private static float getElasticOut(float elapsedTimeRate, double amplitude, double period) {
        if (elapsedTimeRate == 0 || elapsedTimeRate == 1) return elapsedTimeRate;

        double pi2 = Math.PI * 2;
        double s = period / pi2 * Math.asin(1 / amplitude);
        return (float) (amplitude * Math.pow(2, -10 * elapsedTimeRate) * Math.sin((elapsedTimeRate - s) * pi2 / period) + 1);
    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @param amplitude       Amplitude of easing
     * @param period          Animation of period
     * @return easedValue
     */
    private static float getElasticInOut(float elapsedTimeRate, double amplitude, double period) {
        double pi2 = Math.PI * 2;

        double s = period / pi2 * Math.asin(1 / amplitude);
        if ((elapsedTimeRate *= 2) < 1) {
            return (float) (-0.5f * (amplitude * Math.pow(2, 10 * (elapsedTimeRate -= 1f)) * Math.sin((elapsedTimeRate - s) * pi2 / period)));
        }
        return (float) (amplitude * Math.pow(2, -10 * (elapsedTimeRate -= 1)) * Math.sin((elapsedTimeRate - s) * pi2 / period) * 0.5 + 1);

    }
}

File: RxUI/src/main/java/com/tamsiree/rxui/view/likeview/tools/ei/RxEasingProvider.kt
package com.tamsiree.rxui.view.likeview.tools.ei

/**
 * @author tamsiree
 * The Easing class provides a collection of ease functions. It does not use the standard 4 param
 * ease signature. Instead it uses a single param which indicates the current linear ratio (0 to 1) of the tween.
 */
internal object RxEasingProvider {
    /**
     * @param rxEase          Easing type
     * @param elapsedTimeRate Elapsed time / Total time
     * @return easedValue
     */
    operator fun get(rxEase: RxEase, elapsedTimeRate: Float): Float {
        var elapsedTimeRate = elapsedTimeRate
        return when (rxEase) {
            RxEase.LINEAR -> elapsedTimeRate
            RxEase.QUAD_IN -> getPowIn(elapsedTimeRate, 2.0)
            RxEase.QUAD_OUT -> getPowOut(elapsedTimeRate, 2.0)
            RxEase.QUAD_IN_OUT -> getPowInOut(elapsedTimeRate, 2.0)
            RxEase.CUBIC_IN -> getPowIn(elapsedTimeRate, 3.0)
            RxEase.CUBIC_OUT -> getPowOut(elapsedTimeRate, 3.0)
            RxEase.CUBIC_IN_OUT -> getPowInOut(elapsedTimeRate, 3.0)
            RxEase.QUART_IN -> getPowIn(elapsedTimeRate, 4.0)
            RxEase.QUART_OUT -> getPowOut(elapsedTimeRate, 4.0)
            RxEase.QUART_IN_OUT -> getPowInOut(elapsedTimeRate, 4.0)
            RxEase.QUINT_IN -> getPowIn(elapsedTimeRate, 5.0)
            RxEase.QUINT_OUT -> getPowOut(elapsedTimeRate, 5.0)
            RxEase.QUINT_IN_OUT -> getPowInOut(elapsedTimeRate, 5.0)
            RxEase.SINE_IN -> (1f - Math.cos(elapsedTimeRate * Math.PI / 2f)).toFloat()
            RxEase.SINE_OUT -> Math.sin(elapsedTimeRate * Math.PI / 2f).toFloat()
            RxEase.SINE_IN_OUT -> (-0.5f * (Math.cos(Math.PI * elapsedTimeRate) - 1f)).toFloat()
            RxEase.BACK_IN -> (elapsedTimeRate * elapsedTimeRate * ((1.7 + 1f) * elapsedTimeRate - 1.7)).toFloat()
            RxEase.BACK_OUT -> (--elapsedTimeRate * elapsedTimeRate * ((1.7 + 1f) * elapsedTimeRate + 1.7) + 1f).toFloat()
            RxEase.BACK_IN_OUT -> getBackInOut(elapsedTimeRate, 1.7f)
            RxEase.CIRC_IN -> (-(Math.sqrt(1f - elapsedTimeRate * elapsedTimeRate.toDouble()) - 1)).toFloat()
            RxEase.CIRC_OUT -> Math.sqrt(1f - --elapsedTimeRate * elapsedTimeRate.toDouble()).toFloat()
            RxEase.CIRC_IN_OUT -> {
                if (2f.let { elapsedTimeRate *= it; elapsedTimeRate } < 1f) {
                    (-0.5f * (Math.sqrt(1f - elapsedTimeRate * elapsedTimeRate.toDouble()) - 1f)).toFloat()
                } else (0.5f * (Math.sqrt(1f - 2f.let { elapsedTimeRate -= it; elapsedTimeRate } * elapsedTimeRate.toDouble()) + 1f)).toFloat()
            }
            RxEase.BOUNCE_IN -> getBounceIn(elapsedTimeRate)
            RxEase.BOUNCE_OUT -> getBounceOut(elapsedTimeRate)
            RxEase.BOUNCE_IN_OUT -> {
                if (elapsedTimeRate < 0.5f) {
                    getBounceIn(elapsedTimeRate * 2f) * 0.5f
                } else getBounceOut(elapsedTimeRate * 2f - 1f) * 0.5f + 0.5f
            }
            RxEase.ELASTIC_IN -> getElasticIn(elapsedTimeRate, 1.0, 0.3)
            RxEase.ELASTIC_OUT -> getElasticOut(elapsedTimeRate, 1.0, 0.3)
            RxEase.ELASTIC_IN_OUT -> getElasticInOut(elapsedTimeRate, 1.0, 0.45)
            else -> elapsedTimeRate
        }
    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @param pow             pow The exponent to use (ex. 3 would return a cubic ease).
     * @return easedValue
     */
    private fun getPowIn(elapsedTimeRate: Float, pow: Double): Float {
        return Math.pow(elapsedTimeRate.toDouble(), pow).toFloat()
    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @param pow             pow The exponent to use (ex. 3 would return a cubic ease).
     * @return easedValue
     */
    private fun getPowOut(elapsedTimeRate: Float, pow: Double): Float {
        return (1.toFloat() - Math.pow(1 - elapsedTimeRate.toDouble(), pow)).toFloat()
    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @param pow             pow The exponent to use (ex. 3 would return a cubic ease).
     * @return easedValue
     */
    private fun getPowInOut(elapsedTimeRate: Float, pow: Double): Float {
        var elapsedTimeRate = elapsedTimeRate
        return if (2.let { elapsedTimeRate *= it; elapsedTimeRate } < 1) {
            (0.5 * Math.pow(elapsedTimeRate.toDouble(), pow)).toFloat()
        } else (1 - 0.5 * Math.abs(Math.pow(2 - elapsedTimeRate.toDouble(), pow))).toFloat()
    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @param amount          amount The strength of the ease.
     * @return easedValue
     */
    private fun getBackInOut(elapsedTimeRate: Float, amount: Float): Float {
        var elapsedTimeRate = elapsedTimeRate
        var amount = amount
        amount *= 1.525f
        return if (2.let { elapsedTimeRate *= it; elapsedTimeRate } < 1) {
            (0.5 * (elapsedTimeRate * elapsedTimeRate * ((amount + 1) * elapsedTimeRate - amount))).toFloat()
        } else (0.5 * (2.let { elapsedTimeRate -= it; elapsedTimeRate } * elapsedTimeRate * ((amount + 1) * elapsedTimeRate + amount) + 2)).toFloat()
    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @return easedValue
     */
    private fun getBounceIn(elapsedTimeRate: Float): Float {
        return 1f - getBounceOut(1f - elapsedTimeRate)
    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @return easedValue
     */
    private fun getBounceOut(elapsedTimeRate: Float): Float {
        var elapsedTimeRate = elapsedTimeRate
        return if (elapsedTimeRate < 1 / 2.75) {
            7.5625f * elapsedTimeRate * elapsedTimeRate
        } else if (elapsedTimeRate < 2 / 2.75) {
            7.5625f * (1.5f / 2.75f.let { elapsedTimeRate -= it; elapsedTimeRate }) * elapsedTimeRate + 0.75f
        } else if (elapsedTimeRate < 2.5 / 2.75) {
            7.5625f * (2.25f / 2.75f.let { elapsedTimeRate -= it; elapsedTimeRate }) * elapsedTimeRate + 0.9375f
        } else {
            7.5625f * (2.625f / 2.75f.let { elapsedTimeRate -= it; elapsedTimeRate }) * elapsedTimeRate + 0.984375f
        }
    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @param amplitude       Amplitude of easing
     * @param period          Animation of period
     * @return easedValue
     */
    private fun getElasticIn(elapsedTimeRate: Float, amplitude: Double, period: Double): Float {
        var elapsedTimeRate = elapsedTimeRate
        if (elapsedTimeRate == 0f || elapsedTimeRate == 1f) return elapsedTimeRate
        val pi2 = Math.PI * 2
        val s = period / pi2 * Math.asin(1 / amplitude)
        return (-(amplitude * Math.pow(2.0, 10f * 1f.let { elapsedTimeRate -= it; elapsedTimeRate }.toDouble()) * Math.sin((elapsedTimeRate - s) * pi2 / period))).toFloat()
    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @param amplitude       Amplitude of easing
     * @param period          Animation of period
     * @return easedValue
     */
    private fun getElasticOut(elapsedTimeRate: Float, amplitude: Double, period: Double): Float {
        if (elapsedTimeRate == 0f || elapsedTimeRate == 1f) return elapsedTimeRate
        val pi2 = Math.PI * 2
        val s = period / pi2 * Math.asin(1 / amplitude)
        return (amplitude * Math.pow(2.0, -10 * elapsedTimeRate.toDouble()) * Math.sin((elapsedTimeRate - s) * pi2 / period) + 1).toFloat()
    }

    /**
     * @param elapsedTimeRate Elapsed time / Total time
     * @param amplitude       Amplitude of easing
     * @param period          Animation of period
     * @return easedValue
     */
    private fun getElasticInOut(elapsedTimeRate: Float, amplitude: Double, period: Double): Float {
        var elapsedTimeRate = elapsedTimeRate
        val pi2 = Math.PI * 2
        val s = period / pi2 * Math.asin(1 / amplitude)
        return if (2.let { elapsedTimeRate *= it; elapsedTimeRate } < 1) {
            (-0.5f * (amplitude * Math.pow(2.0, 10 * 1f.let { elapsedTimeRate -= it; elapsedTimeRate }.toDouble()) * Math.sin((elapsedTimeRate - s) * pi2 / period))).toFloat()
        } else (amplitude * Math.pow(2.0, -10 * 1.let { elapsedTimeRate -= it; elapsedTimeRate }.toDouble()) * Math.sin((elapsedTimeRate - s) * pi2 / period) * 0.5 + 1).toFloat()
    }
}